drv/i2c-devices/src/pca9538.rs

```diff
@@ -0,0 +1,163 @@
+#[derive(Copy, Clone, Debug, PartialEq, FromPrimitive)]
+#[repr(u8)]
+pub enum Polarity {
+    Normal = 0,
+    Inverted = 1,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, FromPrimitive)]
+enum Register {
+    InputPort = 0x00,
+    OutputPort = 0x01,
+    PolarityInversion = 0x02,
+    Configuration = 0x03,
+}
+
+pub struct Pca9538 {
+    device: I2cDevice,
+}
+
+fn read_reg(
+    device: &I2cDevice,
+    register: Register,
+) -> Result<u8, ResponseCode> {
+    device.read_reg(register as u8)
+}
+
+fn write_reg(
+    device: &I2cDevice,
+    register: Register,
+    value: u8,
+) -> Result<(), ResponseCode> {
+    device.write(&[register as u8, value])
+}
+
+impl Pca9538 {
+    pub fn new(device: I2cDevice) -> Self {
+        Self { device }
+    }
+
+    /// Read the state of pins in the `PinSet`. Note that this results reflects
+    /// the polarity configuration of the pins.
+    pub fn read(&self, pins: PinSet) -> Result<u8, ResponseCode> {
+        read_reg(&self.device, Register::InputPort).map(|r| r & pins.0)
+    }
+
+    /// Set the pins in the `PinSet` to low/high based on the given bool value
+    /// of `set`.
+    pub fn set_to(&self, pins: PinSet, set: bool) -> Result<(), ResponseCode> {
+        let outputs = read_reg(&self.device, Register::OutputPort)?;
+        write_reg(
+            &self.device,
+            Register::OutputPort,
+            if set {
+                outputs | pins.0
+            } else {
+                outputs & !pins.0
+            },
+        )
+    }
+
+    /// Set the pins in the `PinSet`.
+    pub fn set(&self, pins: PinSet) -> Result<(), ResponseCode> {
+        self.set_to(pins, true)
+    }
+
+    /// Reset the pins
+    pub fn reset(&self, pins: PinSet) -> Result<(), ResponseCode> {
+        self.set_to(pins, false)
+    }
+
+    /// Configure the pins in the `PinSet` with the given `Mode` and `Polarity`.
+    pub fn set_mode(
+        &self,
+        pins: PinSet,
+        mode: Mode,
+        polarity: Polarity,
+    ) -> Result<(), ResponseCode> {
+        let output_pins = read_reg(&self.device, Register::Configuration)?;
+        let inverted_pins =
+            read_reg(&self.device, Register::PolarityInversion)?;
+
+        write_reg(
+            &self.device,
+            Register::PolarityInversion,
+            match polarity {
+                Polarity::Normal => inverted_pins & !pins.0,
+                Polarity::Inverted => inverted_pins | pins.0,
+            },
+        )?;
+        write_reg(
+            &self.device,
+            Register::Configuration,
+            match mode {
+                Mode::Input => output_pins | pins.0,
+                Mode::Output => output_pins & !pins.0,
+            },
+        )
+    }
+
+    /// Return the polarity of the pins in the `PinSet`.
+    pub fn polarity(&self, pins: PinSet) -> Result<u8, ResponseCode> {
+        read_reg(&self.device, Register::PolarityInversion).map(|v| v & pins.0)
```

Comment:
Nit:
```suggestion
        read_reg(&self.device, Register::PolarityInversion)? & pins.0
```

Reply:
Ah the return type here was something more specific at some point requiring a bit of wrangling. This is indeed the way.